crates/cranelift/src/stack_switching/instructions.rs

```diff
@@ -0,0 +1,1910 @@
+use cranelift_codegen::ir::BlockArg;
+use itertools::{Either, Itertools};
+
+use cranelift_codegen::ir::condcodes::*;
+use cranelift_codegen::ir::types::*;
+use cranelift_codegen::ir::{self, MemFlags};
+use cranelift_codegen::ir::{Block, BlockCall, InstBuilder, JumpTableData};
+use cranelift_frontend::FunctionBuilder;
+use wasmtime_environ::{PtrSize, TagIndex, TypeIndex, WasmResult, WasmValType};
+
+// TODO(frank-emrich) This is the size for x64 Linux. Once we support different
+// platforms for stack switching, must select appropriate value for target.
+pub const CONTROL_CONTEXT_SIZE: usize = 24;
+
+use super::control_effect::ControlEffect;
+use super::fatpointer;
+
+/// This module contains compile-time counterparts to types defined elsewhere.
+pub(crate) mod stack_switching_helpers {
+    use core::marker::PhantomData;
+    use cranelift_codegen::ir;
+    use cranelift_codegen::ir::InstBuilder;
+    use cranelift_codegen::ir::condcodes::IntCC;
+    use cranelift_codegen::ir::types::*;
+    use cranelift_codegen::ir::{StackSlot, StackSlotKind::*};
+    use cranelift_frontend::FunctionBuilder;
+    use std::mem;
+    use wasmtime_environ::PtrSize;
+
+    #[derive(Copy, Clone)]
+    pub struct VMContRef {
+        pub address: ir::Value,
+    }
+
+    #[derive(Copy, Clone)]
+    pub struct VMHostArray<T> {
+        /// Base address of this object, which must be shifted by `offset` below.
+        base: ir::Value,
+
+        /// Adding this (statically) known offset gets us the overall address.
+        offset: i32,
+
+        /// The type parameter T is never used in the fields above. We still
+        /// want to have it for consistency with
+        /// `wasmtime_environ::Vector` and to use it in the associated
+        /// functions.
+        phantom: PhantomData<T>,
+    }
```

Comment:
This is not a `VMHostArray`, it is a pointer to a `VMHostArray`, right? Can we update the name to reflect that? `VMHostArrayPtr` or `VMHostArrayRef` perhaps?

Also, I don't think we need to separate the static offset and dynamic base address because nothing is (for example) determining whether we need to emit explicit bounds checks or can rely on guard pages for these accesses, so we can just have a single `address: ir::Value` field and trust Cranelift's instruction selection choose the best addressing mode it can given the input IR.